src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,267 @@
+
+    return PublicKey(Slice(blob.data(), blob.size()));
+}
+
+ValidatorList::ValidatorList (beast::Journal j)
+    : j_ (j)
+{
+}
+
+boost::optional<std::string>
+ValidatorList::member (RippleAddress const& identity) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    auto ret = ephemeral_.find (publicKey);
+
+    if (ret != ephemeral_.end())
+        return ret->second;
+
+    ret = permanent_.find (publicKey);
+
+    if (ret != permanent_.end())
+        return ret->second;
+
+    return boost::none;
+}
+
+bool
+ValidatorList::trusted (RippleAddress const& identity) const
+{
+    return static_cast<bool> (member(identity));
+}
+
+bool
+ValidatorList::insertEphemeralKey (
+    PublicKey const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    if (permanent_.find (identity) != permanent_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, identity) <<
+            ": ephemeral key exists in permanent table!";
+        return false;
+    }
+
+    return ephemeral_.emplace (identity, comment).second;
+}
+
+bool
+ValidatorList::removeEphemeralKey (
+    PublicKey const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return ephemeral_.erase (identity);
+}
+
+bool
+ValidatorList::insertPermanentKey (
+    RippleAddress const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    if (ephemeral_.find (publicKey) != ephemeral_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, publicKey) <<
+            ": permanent key exists in ephemeral table!";
+        return false;
+    }
+
+    return permanent_.emplace (publicKey, comment).second;
+}
+
+bool
+ValidatorList::removePermanentKey (
+    RippleAddress const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return permanent_.erase (asPublicKey (identity));
+}
+
+void
+ValidatorList::for_each (
+    std::function<void(PublicKey const&, std::string const&, bool)> func) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    for (auto const& v : permanent_)
+        func (v.first, v.second, false);
+    for (auto const& v : ephemeral_)
+        func (v.first, v.second, true);
+}
+
+//------------------------------------------------------------------------------
+static
+std::size_t
+loadValidators (
+    ValidatorList *vl,
+    std::vector<std::string> const& entries,
+    beast::Journal& j)
+{
+    static boost::regex const re (
+        "^"                         // start of line
+        "(?:\\s*)"                  // whitespace (optional)
+        "([a-zA-Z0-9]+)"            // Node identity
+        "(?:\\s*)"                  // whitespace (optional)
+        "(.*)"                      // comment (optional)
+    );
```

Comment:
I was trying to create a group that didn't require trimming a the white space from the end, but you're right, I needed the comment to be at least two characters.

I'd make one tweak to your regex. The `+` in the second group isn't needed (`.*` will eat those chars first). So: `^[[:space:]]*([[:alnum:]]+)(?:[[:space:]]+(.*[^[:space:]])[[:space:]]*)?`

I had one more idea then I think we've spent enough energy on this. We could use `regex_search` instead of `regex_match` and `match.suffix()` then our regex is really simple. Something like:
```
boost::regex re("[[:alnum:]]+");
if (boost::regex_search(s, match, re))
{
    // match[0]  is the key
    // trim_whitespace(match.suffix()) is the comment
}
```